pkg/telemetry/metrics_db_lock.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2023 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package telemetry
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/vmware-tanzu/tanzu-cli/pkg/common"
+
+	"github.com/juju/fslock"
+)
+
+const (
+	LocalTanzuCLIMetricsDBFileLock = ".cli_metrics_db.lock"
+	// DefaultMetricsDBLockTimeout is the default time waiting on the filelock
+	DefaultMetricsDBLockTimeout = 10 * time.Minute
+)
+
+var cliMetricDBLockFile string
+
+// cliMetricDBLock used as a static lock variable that stores fslock
+// This is used for interprocess locking of the tanzu cli metrics DB file
+var cliMetricDBLock *fslock.Lock
+
+// cliMetricDBMutex is used to handle the locking behavior between concurrent calls
+// within the existing process trying to acquire the lock
+var cliMetricDBMutex sync.Mutex
+
+// AcquireTanzuMetricDBLock tries to acquire lock to update tanzu cli metrics DB file with timeout
+func AcquireTanzuMetricDBLock() {
+	var err error
+
+	if cliMetricDBLockFile == "" {
+		cliMetricDBLockFile = filepath.Join(common.DefaultCLITelemetryDir, LocalTanzuCLIMetricsDBFileLock)
+	}
+
+	// using fslock to handle interprocess locking
+	lock, err := getFileLockWithTimeOut(cliMetricDBLockFile, DefaultMetricsDBLockTimeout)
+	if err != nil {
+		panic(fmt.Sprintf("cannot acquire lock for Tanzu CLI metrics DB, reason: %v", err))
```

Comment:
In the rare scenario of a lock error, I am thinking about, whether or not should we panic here or skip the telemetry collection if this happens for any reason.

Reply:
I was thinking about it too, but since it is very corner case, I choose to go with `Panic`. Now after thinking a bit more, it make sense to return error(which would be ignored for telemetry unless user chooses to show the telemetry errors) instead of `panic` especially if the operation takes too long to complete, and CLI shouldn't give the impression that operation is failed due to telemetry. I will update the PR.

Reply:
updated